packages/sdk-rtl/src/paginator.ts

```diff
@@ -0,0 +1,325 @@
+  const re = /<\s*(.*)\s*>;\s*rel="\s*(.*)\s*"\s*/gm
+  const links = linkHeader.split(',')
+  const obj: PageLinks = {}
+  let arrRes
+
+  links.forEach((link) => {
+    link = link.trim()
+    while ((arrRes = re.exec(link))) {
+      const key = arrRes[2].split(' ')[0].trim().toLocaleLowerCase()
+      obj[key] = {
+        url: arrRes[1].trim(),
+        rel: key as PageLinkRel,
+        name: arrRes[2].trim(),
+      }
+    }
+  })
+  return obj
+}
+
+/**
+ * Create an API paginator for an endpoint that returns a Link header
+ * @param sdk implementation of IAPIMethods. Can be full SDK or functional auth session
+ * @param func sdk call that includes a pagination header
+ * @param options transport options override to capture and use in paging requests
+ *
+ * @example
+ * ```ts
+ * const paged = new paginate(sdk, (sdk) => search_dashboards(sdk, { limit: 10 })
+ * const dashboards = paged.items
+ * while (paged.has('next')) {
+ *   await paged.next()
+ *   dashboards.push(paged.items)
+ * }
+ * console.log(`${paged.total} dashboards retrieved`)
+ * ... (some code to list all dashboards )
+ *
+ * ```
+ */
+export async function paginate<TSuccess, TError>(
+  sdk: IAPIMethods,
+  func: PaginateFunc<TSuccess, TError>,
+  options?: Partial<ITransportSettings>
+): Promise<IPaginate<TSuccess, TError>> {
+  return await new Paginator<TSuccess, TError>(sdk, func, options).init()
+}
+
+/**
+ * Pagination support class
+ */
+export class Paginator<TSuccess, TError>
+  implements IPaginate<TSuccess, TError>
+{
+  items: TSuccess = [] as unknown as TSuccess
+  links: PageLinks = {}
+  total = -1
+  offset = -1
+  limit = -1
+
+  private transport: BaseTransport
+
+  /**
+   * Create an API paginator
+   * @param sdk functional AuthSession or full SDK implementation
+   * @param func sdk function to call
+   * @param options transport overrides to use for subsequent requests
+   */
+  constructor(
+    public sdk: IAPIMethods,
+    public func: PaginateFunc<TSuccess, TError>,
+    public options?: Partial<ITransportSettings>
+  ) {
+    this.transport = sdk.authSession.transport as BaseTransport
+  }
+
+  private async rawCatch(func: () => any) {
+    let raw: IRawResponse = {} as IRawResponse
+    const saved = this.transport.observer
+    try {
+      this.transport.observer = (response: IRawResponse) => {
+        raw = response
+        return response
+      }
+      this.items = await sdkOk(func())
+    } finally {
+      this.transport.observer = saved
+    }
+    if (Object.keys(raw).length === 0 || Object.keys(raw.headers).length === 0)
+      throw new Error('No headers were retrieved for pagination')
+    this.parse(raw)
+    return this
+  }
+
+  get page(): number {
+    if (this.limit < 1 || this.offset < 0) return -1
+    const x = this.offset / this.limit + 1
+    return Math.ceil(x)
+  }
+
+  get pages(): number {
+    if (this.total < 1 || this.limit < 1) return -1
+    const x = this.total / this.limit
+    return Math.ceil(x)
+  }
```

Comment:
readonly properties for reporting/UI purposes. A `setPage(page, limit)` function would be feasible to implement once we've got any result back because we can plug in the calculated values for `limit` and `offset` into the `first` rel link url. If `total` is known, we can also reject the request immediately if the page is beyond the number of available results.